polygeist/test/polygeist-opt/licm.mlir

```diff
@@ -300,46 +300,45 @@ func.func @affine_for_hoist5(%arg0: memref<?x!sycl_accessor_1_f32_rw_gb, 4>) {
+  // CHECK-RELAXED-ALIASING:          sycl.constructor @id(%memspacecast, %c64_i64) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1_, 4>, i64)
   // CHECK-RELAXED-ALIASING-NEXT:     affine.for %arg1 = 0 to 10 {
-  // CHECK-RELAXED-ALIASING-NEXT:      %3 = affine.load %alloca[0] : memref<1x!sycl_id_1_>
-  // CHECK-RELAXED-ALIASING-NEXT:    affine.store %3, %alloca_0[0] : memref<1x!sycl_id_1_>
-  // CHECK-RELAXED-ALIASING-NEXT:      %4 = sycl.accessor.subscript %arg0[%alloca_0] {ArgumentTypes = [memref<?x!sycl_accessor_1_f32_rw_gb, 4>, memref<1x!sycl_id_1_>], FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorIfLi1ELNS0_6access4modeE1026ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi1EvEERfNS0_2idILi1EEE, TypeName = @accessor} : (memref<?x!sycl_accessor_1_f32_rw_gb, 4>, memref<1x!sycl_id_1_>) -> memref<?xf32, 4>
-  // CHECK-RELAXED-ALIASING-NEXT:      %5 = affine.load %4[0] : memref<?xf32, 4>
-  // CHECK-RELAXED-ALIASING-NEXT:      %6 = arith.addf %5, %cst : f32
-  // CHECK-RELAXED-ALIASING-NEXT:      affine.store %6, %4[0] : memref<?xf32, 4>
+  // CHECK-RELAXED-ALIASING-NEXT:      %0 = affine.load %alloca[0] : memref<1x!sycl_id_1_>
+  // CHECK-RELAXED-ALIASING-NEXT:      affine.store %0, %alloca_0[0] : memref<1x!sycl_id_1_>
+  // CHECK-RELAXED-ALIASING-NEXT:      %1 = sycl.accessor.subscript %arg0[%alloca_0] {ArgumentTypes = [memref<?x!sycl_accessor_1_f32_rw_gb, 4>, memref<1x!sycl_id_1_>], FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorIfLi1ELNS0_6access4modeE1026ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi1EvEERfNS0_2idILi1EEE, TypeName = @accessor} : (memref<?x!sycl_accessor_1_f32_rw_gb, 4>, memref<1x!sycl_id_1_>) -> memref<?xf32, 4>
+  // CHECK-RELAXED-ALIASING-NEXT:      %2 = affine.load %1[0] : memref<?xf32, 4>
+  // CHECK-RELAXED-ALIASING-NEXT:      %3 = arith.addf %2, %cst : f32
+  // CHECK-RELAXED-ALIASING-NEXT:      affine.store %3, %1[0] : memref<?xf32, 4>
   // CHECK-RELAXED-ALIASING-NEXT:    }
 
   %alloca = memref.alloca() : memref<1x!sycl_id_1>  
   %alloca_0 = memref.alloca() : memref<1x!sycl_id_1>
   %c64_i64 = arith.constant 64 : i64
   %cst = arith.constant 1.000000e+01 : f32
 
-  %0 = "polygeist.memref2pointer"(%alloca) : (memref<1x!sycl_id_1>) -> !llvm.ptr<!sycl_id_1>
-  %1 = llvm.addrspacecast %0 : !llvm.ptr<!sycl_id_1> to !llvm.ptr<!sycl_id_1, 4>
-  %2 = "polygeist.pointer2memref"(%1) : (!llvm.ptr<!sycl_id_1, 4>) -> memref<?x!sycl_id_1, 4>
-  sycl.constructor @id(%2, %c64_i64) {MangledFunctionName = @_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE} : (memref<?x!sycl_id_1, 4>, i64)
+  %0 = memref.cast %alloca : memref<1x!sycl_id_1> to memref<?x!sycl_id_1>
```

Comment:
Why do we need to cast the static shape information away here?

Reply:
We usually generate this cast, but got canonicalized away with `polygeist.memref2pointer`.

Reply:
Does the `sycl.constructor` not accept `memref<1x!sycl_id_1>`, i.e. memrefs with a static size? If so, why is that?

Reply:
It is a strict translation to the associated function call, which that function takes `memref<?x!sycl_id_1, 4>`. Are you suggesting to abstract the address space and memref shape from `sycl.constructor`?

Reply:
I'm mainly interested in the memref shape here. 

From my understanding of the `memref` type (and the `sycl.constructor` operation definition), I do not see a reason why the constructor or the associated function call would not accept a memref with a statically known shape, if they accept a memref with dynamic shape. After all, the static shape also guarantees that there is at least enough space for one element (the element to construct) in the memory buffer referenced by the memref. 

Is this related to the later lowerings expecting the parameters associated with dynamically shaped memrefs?

Reply:
Yup, `@_ZN4sycl3_V12idILi1EEC1ILi1EEENSt9enable_ifIXeqT_Li1EEmE4typeE` expects dynamically shaped memref.

Reply:
Ok, I see, thanks for the explanation!